Remove deprecated code

mattfarina opened the pull request:
This change removes deprecated code (marked for removal in Helm v4).

This also exposes the Helm template engine function map for SDK users. The removed public function constructor was used by SDK users who were trying to reconstruct some of Helms template functions so their custom code can render render templates. With the removal of this public function that no longer works. By exposing this publicly, those use cases can continue.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**: Removing deprecated code while keeping a solution for community user use cases.

**Special notes for your reviewer**: There are three ways to handle this situation.

1. Remove `NewLookupFunction` and no longer expose this for SDK users. Some of the SDK users are using it outside of Helm. Some of them are partially providing Helms functions. I did not find a public app on GitHub using all of Helms template functions. Removing it means they will need to vendor a copy of our code or write their own.
2. Do not deprecate the `NewLookupFunction` function and and instead make it `newLookupFunction`. This will help some community users who use it.
3. Expose all the Helm template functions for the community.

My initial reaction was to go with (1) to keep this code Helm focused. 

If we are going to expose our template functions, we could do so with those in mind who are working with Helm charts and need all our functions. This is to power SDK users who are doing Helm work. That's why I went with (3) in this PR.

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

mattfarina committed:
Remove ValidName regex

This regex was already deprecated.

Validation happens inside the Metadata Validate function for the
name instead of using this regex.

Signed-off-by: Matt Farina <matt.farina@suse.com>

robertsirc reviewed (dismissed):
LGTM

banjoh reviewed (changes requested):
I attempted invoking `FuncMap(...)` but got blocked constructing `*ClientProvider`. For me to get the same behaviour as the SDK, I'd need to reimplement what's in `clientProviderFromConfig`

Two options come to mind to address this
- Have `FuncMap` accept `*rest.Config` instead of `*ClientProvider` just like the deprecated function had
- Expose a constructor for `*ClientProvider` to use when invoking `FuncMap(...)` i.e
```go
func NewClientProvider(config *rest.Config) ClientProvider {
	return clientProviderFromConfig{config: config}
}
```

**Nice to have**: Can we make `lookupFunc` function prototype public? This way, a client does not have to redeclare it

Here is my sample code

```go
type LookupFunc = func(apiversion string, resource string, namespace string, name string) (map[string]interface{}, error)

func lookup(apiversion string, resource string, namespace string, name string) (map[string]interface{}, error) {
	var cfg *rest.Config
	cp := engine.NewClientProvider(cfg)
	lkFunc := engine.FuncMap(false, false, &cp)["lookup"]
	// OR: lkFunc := engine.FuncMap(false, false, cfg)["lookup"]
	lookupFunc, ok := lkFunc.(LookupFunc)
	if !ok {
		return nil, fmt.Errorf("Failed to cast lookup function")
	}

	return lookupFunc(apiversion, resource, namespace, name)
}
```

scottrigby reviewed (changes requested):
Thanks @mattfarina for reaching out about existing use of this function.

To continue allowing end users to make use of this functionality, we would need to either add an additional function as @banjoh said in https://github.com/helm/helm/pull/30686#pullrequestreview-2702482532, or reconsider the earlier (unexplained) comment to deprecate this function and just leave as-is in Helm 4.

mattfarina committed:
Removing deprecation notice for this function.

While the constructor is not used by Helm itself, it is used by SDK
users and there is currently no alternative way to expose this.

Signed-off-by: Matt Farina <matt.farina@suse.com>

mattfarina commented:
@scottrigby @banjoh  can you please re-review this. I removed the deprecated notice on the public function in the SDK that y'all are using. This is the simpler path.

banjoh reviewed (approved):
LGTM

scottrigby reviewed (approved):
Lgtm 👍
